### PR TITLE
feat(dts-generator): trust source data for event parameter optionality

### PR DESCRIPTION
**Pull request by akudev**
Previously, all properties in generated $Xxx$YyyEventParameters interfaces were unconditionally forced to optional (?), regardless of their actual "optional" value in the api.json source data.

Now the generator respects each parameter's optional field by default:
- optional: false → required (no ?)
- optional: true or undefined → optional (?)

For events whose parameter optionality has not yet been verified in the OpenUI5/SAPUI5 sources, a new "eventsWithAllParamsOptional" directive (string[] in .dtsgenrc) preserves the legacy all-optional behavior. Format: "fully.qualified.ClassName:eventName"

Changes:

- generate-from-objects.ts: Add eventsWithAllParamsOptional to Directives.

- generate.ts: Initialize the new directive field.

- json-fixer.ts: Pass directives to addEventParameterInterfaces(). Add "events" to mapLikeProperties so .dtsgenrc overlays can target individual events by name.

- json-event-parameter-interfaces.ts: Invert buildProperties() to respect source data by default; only force all-optional when the event is in the eventsWithAllParamsOptional deny-list. Log a warning when a parameter has no "optional" value (unexpected edge case).

- Regenerated snapshot .d.ts files (parameters with optional: false now appear as required).

Needed for https://github.com/UI5/typescript/issues/529

### CODE REVIEW

File: test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+        !selectedItemLine.includes("?"),
+        `selectedItem should be required (no '?'), got: "${selectedItemLine.trim()}"`,
+      );
+      assert.ok(
+        valueLine.includes("?"),
+        `value should be optional (has '?'), got: "${valueLine.trim()}"`,
+      );
+    });
+
+    it("parameters without optional field default to optional", async () => {
```
Comment (akudev) [inline test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js:265]: this shouldn't happen, as all parameters should have optional true or false

File: packages/dts-generator/src/phases/json-fixer.ts [hunk 2]
```diff
@@ -49,7 +49,7 @@ function mergeOverlays(
     );
   }
 
-  const mapLikeProperties = new Set(["methods", "properties"]);
+  const mapLikeProperties = new Set(["methods", "properties", "events"]);
```
Comment (akudev) [inline packages/dts-generator/src/phases/json-fixer.ts:52]: Not strictly needed, but allows also .dtsgenrc overlays for events

File: packages/dts-generator/src/utils/json-event-parameter-interfaces.ts [hunk 3]
```diff
@@ -71,20 +73,32 @@ function createEventParameterInterfaces(
...
+        prop,
       );
+      if (!allOptional) {
+        if (prop.optional === false) {
+          transformed.optional = false;
+        } else if (prop.optional !== true) {
+          log.warn(
+            `Event parameter "${prop.name}" in ${context} has no "optional" value; treating as optional.`,
+          );
+        }
```
[the remark this answers is not in the document]
Comment (akudev) [reply packages/dts-generator/src/utils/json-event-parameter-interfaces.ts:100]: done

File: packages/dts-generator/src/generate-from-objects.ts [hunk 4]
```diff
@@ -103,6 +103,13 @@ export interface Directives {
...
    */
   modulesWithNamedExports: string[];
+
+  /**
+   * Events listed here retain the legacy "all parameters optional" behavior.
+   * Format: "fully.qualified.ClassName:eventName"
+   * Used for events whose parameter optionality has not yet been verified in the source.
+   */
+  eventsWithAllParamsOptional?: string[];
 }
```
[the remark this answers is not in the document]
Comment (akudev) [reply packages/dts-generator/src/generate-from-objects.ts:113]: done

File: test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+
+      // The child interface should extend the parent interface
+      assert.match(
+        result.dtsText,
+        /Child\$ChangeEventParameters extends Parent\$ChangeEventParameters/,
+      );
+    });
+
+    it("parent deny-listed: inherited params stay optional, child's new params are required", async () => {
+      const childEvent = makeEvent("change", {
```
Comment (codeworrior) [inline test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js:768]: Logically, if a Child documents that the inherited event parameter `value` is mandatory, shouldn't it be mandatory in the child's event parameters? IMO, this would be well embedded into the parents API. The opposite would be strange: when a child makes a parameter optional that was mandatory in the parent. This could break code that only deals with the parent API.

Or did you think about cases where, even for the subclass, the parent code still fires an event, e.g. w/o the parameter? Valid scenario. But then I think it would be a mistake for the Child to document the parameter as mandatory.

Oh, sorry, I was reading the tests bottom-up. You don't redeclare inherited event parameters in the child's event parameter type. Makes sense. Then, the optionality obviously can't change in the subtypes.

Ignore my comment.

File: packages/dts-generator/src/utils/json-event-parameter-interfaces.ts [hunk 6]
```diff
@@ -71,20 +74,32 @@ function createEventParameterInterfaces(
...
+      const transformed = addJsDocProps(
+        {
+          name: prop.name,
+          type: prop.type,
+          visibility: "public", // prop.visibility,
+        },
+        prop,
       );
+      if (!allOptional) {
+        if (prop.optional === false) {
```
Comment (codeworrior) [inline packages/dts-generator/src/utils/json-event-parameter-interfaces.ts:94]: IIRC, you voted for a default value of "false" for such boolean properties...

File: test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js [hunk 7]
```diff
@@ -0,0 +1,863 @@
...
+      );
+      assert.ok(
+        valueLine.includes("?"),
+        `value should be optional (has '?'), got: "${valueLine.trim()}"`,
+      );
+    });
+
+    // All parameters should have optional: true or false in practice;
+    // this test documents the defensive fallback behavior (treated as optional).
+    it("parameters without optional field default to optional", async () => {
```
Comment (codeworrior) [inline test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js:265]: I didn't expect this default, see above.

**Review by codeworrior — DISMISSED** [review]
The default for `optional` surprises me, but as it is always set, I don't really mind. Besides that, the change looks reasonable to me.

File: packages/dts-generator/src/utils/json-event-parameter-interfaces.ts (hunk 6 above)
Comment (akudev) [reply packages/dts-generator/src/utils/json-event-parameter-interfaces.ts:94]: hm, did I? As you wrote below, it doesn't matter as this case shouldn't happen anyway, but optional==true as fallback means we can't guarantee that there is always a value, so it sounds safe as default.

File: test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js (hunk 7 above)
Comment (akudev) [reply test-packages/openui5-snapshot-test/test/event-params-optionality-spec.js:265]: resolving because this is a precondition for merging
